pkg/operator/agent_manager.go

```diff
@@ -1,14 +1,221 @@
+	"k8s.io/client-go/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/pkg/labels"
+)
+
+var (
+	managedByOperatorLabels = map[string]string{
+		"managed-by": "container-linux-update-operator",
+		"app":        agentDefaultAppName,
+	}
+
+	daemonsetName = "container-linux-update-agent-ds"
+)
 
 // updateAgent updates the agent on nodes if necessary.
-// It returns a value indicating whethre the update-agent is currently up to
-// date.
-func (k *Kontroller) updateAgent(nodeList *v1.NodeList) (bool, error) {
-	agentDaemonset, err := k.kc.DaemonSets(k.namespace).List(&v1.ListOptions{})
+//
+// NOTE: the version for the agent is assumed to match the versioning scheme
+// for the operator, thus our version is used to figure out the appropriate
+// agent version.
+// Furthermore, it's assumed that all future agent versions will be backwards
+// compatible, so if the agent's version is greater than ours, it's okay.
+func (k *Kontroller) runDaemonsetUpdate(agentImageRepo string) error {
+	agentDaemonsets, err := k.kc.DaemonSets(k.namespace).List(v1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set(managedByOperatorLabels)).String(),
+	})
+	if err != nil {
+		return err
+	}
+
+	if len(agentDaemonsets.Items) == 0 {
+		// No daemonset, create it
+		runErr := k.createAgentDamonset(agentImageRepo)
+		if runErr != nil {
+			return runErr
+		}
+		// runAgent succeeded, all should be well and converging now
+		return nil
+	}
+
+	// There should only be one daemonset since we use a well-known name and
+	// patch it each time rather than creating new ones.
+	if len(agentDaemonsets.Items) > 1 {
+		glog.Errorf("only expected one daemonset managed by operator; found %+v", agentDaemonsets.Items)
+		return fmt.Errorf("only expected one daemonset managed by operator; found %v", len(agentDaemonsets.Items))
+	}
+
+	agentDS := agentDaemonsets.Items[0]
+
+	var dsSemver semver.Version
+	if dsVersion, ok := agentDS.Annotations[constants.AgentVersion]; ok {
+		ver, err := semver.Parse(dsVersion)
+		if err != nil {
+			return fmt.Errorf("agent daemonset had version annotation, but it was not valid semver: %v[%v] = %v", agentDS.Name, constants.AgentVersion, dsVersion)
+		}
+		dsSemver = ver
+	} else {
+		glog.Errorf("managed daemonset did not have a version annotation: %+v", agentDS)
+		return fmt.Errorf("managed daemonset did not have a version annotation")
+	}
+
+	if dsSemver.LT(version.Semver) {
+		// daemonset is too old, update it
+		// TODO: perform a proper rolling update rather than delete-then-recreate
+		// Right now, daemonset rolling updates aren't upstream and are thus fairly
+		// painful to do correctly. In addition, doing it correctly doesn't add too
+		// much value unless we have corresponding detection/rollback logic.
+		falseVal := false
+		err := k.kc.DaemonSets(k.namespace).Delete(agentDS.Name, &v1.DeleteOptions{
+			OrphanDependents: &falseVal, // Cascading delete
+		})
+		// TODO: this requires delete to be blocking to work. Is it?
```

Comment:
In practice it always succeeds, so good enough for me to delete that comment.

The failure case is it restarting and trying again, so I'll just remove the comment.